discord-scripts/channel-management.ts

```diff
@@ -0,0 +1,185 @@
+  const fetchedMessages = await channel.messages.fetch(options)
+
+  if (fetchedMessages.size === 0) {
+    return new Collection<string, Message<true>>()
+  }
+
+  const lastId = fetchedMessages.lastKey()
+  const olderMessages = await fetchAllMessages(channel, lastId)
+
+  return new Collection<string, Message<true>>().concat(
+    fetchedMessages,
+    olderMessages,
+  )
+}
+
+export default async function archiveChannel(
+  discordClient: Client,
+  robot: Robot,
+) {
+  const { application } = discordClient
+
+  if (application) {
+    // check if archive-channel command already exists, if not create it
+    const existingArchiveCommand = (await application.commands.fetch()).find(
+      (command) => command.name === "archive-channel",
+    )
+    if (existingArchiveCommand === undefined) {
+      robot.logger.info("No archive-channel command found, creating it!")
+      await application.commands.create({
+        name: "archive-channel",
+        description:
+          "Archives channel to archived channels category (Defense only)",
+      })
+      robot.logger.info("archive channel command set")
+    }
+
+    // check if unarchive-channel command already exists, if not create it
+    const existingUnarchiveCommand = (await application.commands.fetch()).find(
+      (command) => command.name === "unarchive-channel",
+    )
+    if (existingUnarchiveCommand === undefined) {
+      robot.logger.info("No unarchive-channel command found, creating it!")
+      await application.commands.create({
+        name: "unarchive-channel",
+        description:
+          "unarchive channel back to defense category (Defense only)",
+      })
+      robot.logger.info("unarchive channel command set")
+    }
+
+    // move channel to archived-channel category and send out transcript to interaction
+    discordClient.on("interactionCreate", async (interaction) => {
+      if (
+        !interaction.isCommand() ||
+        interaction.commandName !== "archive-channel"
+      ) {
+        return
+      }
+      if (!interaction.guild || !(interaction.channel instanceof TextChannel)) {
+        return
+      }
+
+      try {
+        const allMessages = await fetchAllMessages(interaction.channel)
+        robot.logger.info(`Total messages fetched: ${allMessages}`)
+
+        const messagesContent = allMessages
+          .reverse()
+          .map(
+            (m) =>
+              `${m.createdAt.toLocaleString()}: ${m.author.username}: ${
+                m.content
+              }`,
+          )
+          .join("\n")
+
+        const filePath = `${interaction.channel.name}_transcript.txt`
+        await writeFile(filePath, messagesContent, "utf-8")
+
+        // check for or create archived category
+        let archivedCategory = interaction.guild.channels.cache.find(
+          (c) =>
+            c.type === ChannelType.GuildCategory &&
+            c.name.toLowerCase() === "archived-channels",
```

Comment:
As a category name, this one will look weird, since categories are all caps and can have any characters. We currently have a category `Archive: Defense` that is the correct fit here.